pkg/await/batch_job.go

```diff
@@ -0,0 +1,211 @@
+// Pod fails, (e.g., volume fails to mount, image fails to pull, exited with code 1, etc.).
+//
+//
+// x-refs:
+//   * https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
+
+// --------------------------------------------------------------------------
+
+const (
+	DefaultJobTimeoutMins = 10
+)
+
+type jobInitAwaiter struct {
+	job      *unstructured.Unstructured
+	config   createAwaitConfig
+	state    *states.StateChecker
+	errors   logging.TimeOrderedLogSet
+	resource ResourceId
+}
+
+func makeJobInitAwaiter(c createAwaitConfig) *jobInitAwaiter {
+	return &jobInitAwaiter{
+		config:   c,
+		job:      c.currentOutputs,
+		state:    states.NewJobChecker(),
+		resource: ResourceIdFromUnstructured(c.currentOutputs),
+	}
+}
+
+func (jia *jobInitAwaiter) Await() error {
+	jobClient, err := clients.ResourceClient(kinds.Job, jia.config.currentInputs.GetNamespace(), jia.config.clientSet)
+	if err != nil {
+		return errors.Wrapf(err,
+			"Could not make client to watch Job %q",
+			jia.config.currentInputs.GetName())
+	}
+	jobWatcher, err := jobClient.Watch(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "Couldn't set up watch for Job object %q",
+			jia.config.currentInputs.GetName())
+	}
+	defer jobWatcher.Stop()
+
+	podAggregator, err := NewPodAggregator(ResourceIdFromUnstructured(jia.job), jia.config.clientSet)
+	if err != nil {
+		return errors.Wrapf(err, "Could not create PodAggregator for %s", jia.resource.GVKString())
+	}
+	defer podAggregator.Stop()
+
+	timeout := metadata.TimeoutDuration(jia.config.timeout, jia.config.currentInputs, DefaultJobTimeoutMins*60)
+	for {
+		if jia.state.Ready() {
+			return nil
+		}
+
+		// Else, wait for updates.
+		select {
+		case <-jia.config.ctx.Done():
+			return &cancellationError{
+				object:    jia.job,
+				subErrors: jia.errorMessages(),
+			}
+		case <-time.After(timeout):
+			return &timeoutError{
+				object:    jia.job,
+				subErrors: jia.errorMessages(),
+			}
+		case event := <-jobWatcher.ResultChan():
+			err := jia.processJobEvent(event)
+			if err != nil {
+				return err
+			}
+		case messages := <-podAggregator.ResultChan():
+			for _, message := range messages {
+				jia.errors.Add(message)
+				jia.config.logMessage(message)
+			}
+		}
+	}
+}
+
+func (jia *jobInitAwaiter) Read() error {
+	jobClient, err := clients.ResourceClient(kinds.Job, jia.config.currentInputs.GetNamespace(), jia.config.clientSet)
+	if err != nil {
+		return errors.Wrapf(err,
+			"Could not make client to get Job %q",
+			jia.config.currentInputs.GetName())
+	}
+	// Get live version of Job.
+	job, err := jobClient.Get(jia.config.currentInputs.GetName(), metav1.GetOptions{})
+	if err != nil {
+		// IMPORTANT: Do not wrap this error! If this is a 404, the provider need to know so that it
+		// can mark the Pod as having been deleted.
+		return err
+	}
+
+	_ = jia.processJobEvent(watchAddedEvent(job))
```

Comment:
Just for my own education: you _don't_ have to try to get the pods here? In state of error, there's probably a lot of useful information there, and they'll be around (I think?) until the job is complete. In state of success the pods are gone, but the job reports success? Am I missing something?

Reply:
See https://github.com/pulumi/pulumi-kubernetes/issues/823